### PR TITLE
Create SQL logic for CreateNotificationAction and add relevant tests for v9 migration

### PR DESCRIPTION
**Pull request by hhdqirui**
Part of #12048 

Create SQL logic for `CreateNotificationAction` and add relevant tests.

### CODE REVIEW

File: src/test/java/teammates/storage/sqlapi/NotificationsDbTest.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+        when(session.get(Notification.class, notificationId.toString())).thenReturn(null);
+
+        notificationsDb.createNotification(newNotification);
+
+        verify(session, times(1)).persist(newNotification);
+    }
+
+    @Test
+    public void testCreateNotification_invalidNonNullParameters() {
+        ______TS("failure: start time is before end time");
```
[the remark this answers is not in the document]
Comment (samuelfangjw) [reply src/test/java/teammates/storage/sqlapi/NotificationsDbTest.java:62]: In general, for unit tests we really want relatively short tests (~5 lines as a guideline) that can be written really quickly that test a single behaviour/aspect.

File: src/main/java/teammates/sqllogic/core/NotificationsLogic.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+import teammates.storage.sqlentity.Notification;
+
+/**
+ * Handles the logic related to notifications.
+ */
+public final class NotificationsLogic {
+
+    private static final NotificationsLogic instance = new NotificationsLogic();
+
+    private final NotificationsDb notificationsDb = NotificationsDb.inst();
```
Comment (samuelfangjw) [inline src/main/java/teammates/sqllogic/core/NotificationsLogic.java:15]: Pass in notificationsDb as a dependency instead (via initLogicDependencies). This way, it's easier to mock this out if we need to do unit testing (for the more complex logic methods, once we move the logic out of the db layer).

File: src/it/java/teammates/it/storage/sqlapi/NotificationDbIT.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+                .withStartTime(Instant.parse("2011-01-01T00:00:00Z"))
+                .withEndTime(Instant.parse("2099-01-01T00:00:00Z"))
+                .withStyle(NotificationStyle.DANGER)
+                .withTargetUser(NotificationTargetUser.GENERAL)
+                .withMessage("<p>Deprecation happens in three minutes</p>")
+                .build();
+
+        assertThrows(InvalidParametersException.class, () -> notificationsDb.createNotification(invalidNotification2));
+
+        ______TS("failure: empty message");
```
Comment (samuelfangjw) [inline src/it/java/teammates/it/storage/sqlapi/NotificationDbIT.java:77]: For integration tests, try not to test too much. There's quite significant costs to each additional test (running time, dev time, maintainability, etc.). Especially since in this context the devs are the ones that maintain both integration & unit tests, we can skip some redundant tests.

Not a "law", but in general, we want to avoid testing things that do not test the integration between modules/components and are already covered by our unit test. Let the unit tests, which have much less associated costs, check for these (usually edge) cases. In this case, we only want to test the integration between db layer and db/hibernate itself, so we can consider not testing for edge cases that do not interact with the db.

**Review by samuelfangjw — CHANGES_REQUESTED** [review]

File: src/main/java/teammates/storage/sqlapi/NotificationsDb.java [hunk 4]
```diff
@@ -35,7 +37,7 @@ public Notification createNotification(Notification notification)
             throw new InvalidParametersException(notification.getInvalidityInfo());
         }
 
-        if (getNotification(notification.getNotificationId().toString()) != null) {
+        if (notification.getNotificationId() != null && getNotification(notification.getNotificationId()) != null) {
             throw new EntityAlreadyExistsException(String.format(ERROR_CREATE_ENTITY_ALREADY_EXISTS,
```
Comment (jianhandev) [inline src/main/java/teammates/storage/sqlapi/NotificationsDb.java:41]: At this point I think `notification.getNotificationId()` will always return `null` as the ID generation only happens when session is flushed, which means this block of code will never be run at all.

**Review by samuelfangjw — APPROVED** [review]
